New movement!

Bod9001 opened the pull request:
### Purpose
to reduce lag to, reduce rubber banding, 
Less spaghetti

generally all-around improvements for movement

### Notes:
StopRandomMidGameCompile_import.cs
I've added this to stop the game compiling while playing ,  note it does pop up a asset importing window,
 it doesn't do anything though You can just move it out of the way and play the game normally  in editor

TODO 
other player movement looks funny ( it gets two messages, set target and force push  it should use set target but it's a pain to work out which is which)

Characters pushing against each other is a bit buggy (even with locker line)
reset doesn't set register tile sometimes?
movement desynchronising  with constant movement ( when it is constantly reset idk if can do this  ) 

rethink space wind pushing player

moving across  matrixs if swap to quickly between multiple , causes appearing in space 
( it's to do with registered tile matrix ID, don't feel like doing it in this PR)

### Changelog:

CL: [New] players can be thrown
CL: [Improvement] Rubber banding should be less severe
CL: [Improvement] server performance should be better

Bod9001 committed:
Beginnings of new movement

Bod9001 committed:
Better Formatting

NoooneyDude commented on UnityProject/Assets/Scripts/Player/PlayerAction.cs:
```diff
@@ -22,14 +31,55 @@ public Vector2Int Direction()
+			return MovementSynchronisation.PlayerMoveDirection.Up_Right;
+		}
+
+		return MovementSynchronisation.PlayerMoveDirection.Up;
```
Probably better to use a static dictionary.

Bod9001 replied:
ok will do

Bod9001 committed:
Slowly adding more stuff

Bod9001 committed:
Need to think  thrrough pushing prediiction more

Bod9001 committed:
Todaay''s wwork

Bod9001 committed:
Moveement objects

Bod9001 committed:
Should detect if slipping

MaxIsJoe commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/ObjectLayer.cs:
```diff
@@ -98,19 +159,61 @@ public float GetObjectResistanceAt( Vector3Int position, bool isServer )
+		bool CanPushObjects = false;
+
+		//Targeting windoors here
+		foreach (RegisterTile o in GetTileList(isServer).Get(to))
```
This seems like it's the same foreach loop in 106, Can't you turn this into a function to reduce redundancy?

Bod9001 replied:
well is quite small, Can do later

Bod9001 committed:
Update

Bod9001 committed:
Add pushing and bumping Interactions

Bod9001 committed:
Adds gravity and everything has a size now

Bod9001 committed:
Universal physics now has  Space movement

Bod9001 committed:
Updates

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
Improvements

Bod9001 committed:
Improvements

Bod9001 committed:
Changes

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
Fixed a few things

In terms of
Testing out the functionality
Apart from networking,
A make it so items being pulled in space, Follow the path of the player,
B Check performance from removing and adding from that list

Bod9001 committed:
Pulling improvements

Bod9001 committed:
Walking aaroundd on moving matrices  noww works better

Bod9001 committed:
Starting networking

Bod9001 committed:
Networkinng

Bod9001 committed:
Networking impprovements

Bod9001 committed:
Code updates

Bod9001 committed:
Fixes movement and, Mid rotated chattels should handle Ping better

Bod9001 committed:
code

Bod9001 committed:
No more CNT

Bod9001 committed:
No more pushpull!

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
Hopefully the Correct changes

Bod9001 committed:
Map chhanges

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
Prefabs and changes

Bod9001 committed:
Prefab changes and code

Bod9001 committed:
Bouncy

Bod9001 committed:
Code changess

Bod9001 committed:
More improvements

Bod9001 committed:
Wonder what I missed

Bod9001 committed:
Banana peel

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
Fixes merge

Bod9001 committed:
all the prefabs that need to be parented properly

Bod9001 committed:
Improvements

Bod9001 committed:
a TODO

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
Code improvements

Bod9001 committed:
Slipping and pulling improvements

Bod9001 committed:
Adds RCS control

Bod9001 committed:
Messing around with objects equalisation

Bod9001 committed:
Update Missing things

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
Changees

Bod9001 committed:
Merge branch 'develop' into new-move

Bod9001 committed:
Footsteps

Bod9001 committed:
Bug fixes and improvemmentts

Bod9001 committed:
Bunch of fixes and improvements

Bod9001 committed:
makes it so you cancel pulling if you walk into the object you're pulling

Bod9001 committed:
Improvements

MaxIsJoe committed:
Syncs intent and Adds walk mode change

MaxIsJoe committed:
Update MovementSynchronisation.cs

Bod9001 committed:
Synchronisse intent

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
Improved slipping

Bod9001 committed:
Prefab changes, hopefully got everything

Bod9001 committed:
No admin UI Staying open

Bod9001 committed:
Fixes infinite loops

Bod9001 committed:
Changes that were accidentally included

Bod9001 committed:
Lobby fixes

and it is pushble

Bod9001 committed:
fixes loop and fixes admin UI on scene launch directtly

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

MaxIsJoe committed:
Merge branch 'new-move' into g

Bod9001 committed:
Merge pull request #3 from MaxIsJoe/g

Intent sync and re-added back the ability to change speeds

Bod9001 committed:
space wind!

Bod9001 committed:
fixes not getting onto rotated shuttles

Bod9001 committed:
fixes lockers being pushed into walls

Bod9001 committed:
remove grandma's lag spikes from bad package code

Bod9001 committed:
Small balance

Bod9001 committed:
fixes tests

Bod9001 committed:
Merge branch 'develop' into new-move

Bod9001 committed:
fix compile

corp-0 commented on UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs:
```diff
@@ -0,0 +1,1631 @@
+
+public class UniversalObjectPhysics : NetworkBehaviour, IRightClickable
+{
+	//TODO Push objects being Reset our client
```
are all of these TODO still TODO?

Bod9001 replied:
some of them aren't

corp-0 replied:
it is probably a better idea to move them into issues (each one), after we merged this

corp-0 commented on UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs:
```diff
@@ -0,0 +1,1631 @@
+
+	public void StoreTo(ObjectContainer NewParent)
+	{
+		if (NewParent == this) return; //Storing something inside of itself what?
```
spawn a blackhole, of course

Bod9001 replied:
hehe

NoooneyDude replied:
You should totally do this 😂 give some developer a laugh if they make a mistake.
And a serverwide message "Something has defeated physics! An ominous feeling washes over you..." so the reason why is easy to find

corp-0 commented on UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs:
```diff
@@ -0,0 +1,1631 @@
+
+
+	[RightClickMethod()]
+	public void ThrowWithSlide()
```
are these admin only?

Bod9001 replied:
there meant to be debug only , Easy enough to clean up

corp-0 commented on UnityProject/Assets/Scripts/Util/RandomUtils.cs:
```diff
@@ -4,6 +4,14 @@
 using UnityEngine;
 using Random = UnityEngine.Random;
 
+public enum SpinMode
```
feels weird to have this here- It is RandomUtils as in utilities for handling randomization

Bod9001 replied:
where would be the best place to put it?

corp-0 replied:
somewhere closer to where it is used, which I assume is the UniversalPhysics thingy. Maybe create a new file for all enums it uses? Like we do for health.

Bod9001 replied:
could do

NoooneyDude commented on UnityProject/Assets/Scripts/Managers/UpdateManager/ManagedBehaviour.cs:
```diff
@@ -21,7 +21,7 @@ protected virtual void OnDisable()
 	///     Also don't forget to call ManagedNetworkBehaviour.Awake(); first.
 	///     If your class does not use the Awake function, this object will be added to the UpdateManager automatically.
-	///     Do not forget to replace your Update function with public override void UpdateMe()
+	///     Do not forget to replace your Update function with public override void FlyingUpdateMe()
```
`FlyingUpdateMe()`?

Ah I see, global rename or something. There's a few other scripts affected.

Bod9001 replied:
yeah, when Ryder doesn't do a good job

NoooneyDude commented on UnityProject/Assets/Scripts/Messages/Client/RequestMoveMessage.cs:
```diff
@@ -15,7 +15,7 @@ public struct NetMessage : NetworkMessage
 		public override void Process(NetMessage msg)
 		{
-			SentByPlayer.Script.PlayerSync.ProcessAction(msg.Action);
+			//SentByPlayer.Script.PlayerSync.ProcessAction(msg.Action);
```
So we can remove this script?

Bod9001 replied:
should be able to yeah

NoooneyDude commented on UnityProject/Assets/Scripts/Messages/Server/InformPullMessage.cs:
```diff
@@ -1,92 +1,2 @@
 using System.Collections.Generic;
 using Mirror;
-
```
Can remove this script too I suppose.

NoooneyDude commented on UnityProject/Assets/Scripts/Objects/Construction/Girder.cs:
```diff
@@ -182,15 +182,15 @@ public void ServerPerformInteraction(HandApply interaction)
 		public string Examine(Vector3 worldPos)
 		{
-			return (objectBehaviour.IsPushable ? "Use a wrench to secure to the floor, or a screwdriver to disassemble it."
+			return (objectBehaviour.IsNotPushable! ? "Use a wrench to secure to the floor, or a screwdriver to disassemble it."
```
Exclamation mark at the end? Null forgiving operator... interesting. Though I guess this isn't what we want here.

Bod9001 replied:
== false right?

corp-0 replied:
in plain English, what do you want to return there?

Bod9001 replied:
it should be the opposite of objectBehaviour.IsPushable  would be

Bod9001 replied:
because it's IsNotPushable

NoooneyDude replied:
Yep `== false` sounds good

NoooneyDude commented on UnityProject/Assets/Scripts/Objects/Gateway/TransportUtility.cs:
```diff
@@ -78,7 +68,7 @@ public static void TransportObjectAndPulled(PushPull pushPullObject, Vector3 tra
 
 				//Disconnect pulling to make it not be a problem
-				currentObj.ServerStopPulling();
+				currentObj.PullSet(null, false); //TODO Test without
```
TODO here

NoooneyDude commented on UnityProject/Assets/Scripts/Objects/ObjectContainer.cs:
```diff
@@ -135,13 +165,13 @@ public void StoreObjects(IEnumerable<GameObject> gameObjects)
 
 				// Can't store secured objects (exclude this check on mobs as e.g. magboots set pushable false)
-				if (entity.TryGetComponent<HealthV2.LivingHealthMasterBase>(out _) == false && entity.IsPushable == false) continue;
+				if (entity.IsNotPushable) continue;
```
Why remove this check? As far as I'm aware, the comment above still applies.

Bod9001 replied:
I removed the player specific storage, and is now just one storage

NoooneyDude replied:
Hmm I'm not sure what you mean. So the original issue was that players with magboots enabled (which set the player as immovable) couldn't enter lockers - they'd stay outside when the locker closed.

Tbh the magboots shouldn't be setting the player as immovable but be doing something else.

Bod9001 replied:
yeah, I think going for that approach is good and I can do it as part of rethinking space wind, I'll do that in a separate PR though since it's just a generic push currently

NoooneyDude commented on UnityProject/Assets/Scripts/Objects/Other/PortableFlasher.cs:
```diff
@@ -49,7 +49,7 @@ private void WrenchInteraction(PlayerScript wrenchHolder)
 			isWrenched = !isWrenched;
 			//We inverse this to get the opposite of the wrench, so if its not wrenched; isPushable is true and vice versa
-			objectBehaviour.ServerSetPushable(!isWrenched);
+			objectBehaviour.SetIsNotPushable(!isWrenched);
```
```suggestion
			objectBehaviour.SetIsNotPushable(isWrenched);
```

Bod9001 replied:
isn't the point that it changes each time you interact with it?

NoooneyDude replied:
So the issue is that later on the examine message is wrong, it is now inverted.

Bod9001 replied:
done

NoooneyDude commented on UnityProject/Assets/Scripts/Player/PlayerScript.cs:
```diff
@@ -380,7 +381,7 @@ public void SyncPlayerName(string oldValue, string value)
 	}
 
-	public bool IsHidden => !PlayerSync.ClientState.Active;
+	public bool IsHidden => false; //TODO
```
TODO - does anything use this?

Bod9001 replied:
oh yeah, need to do that

NoooneyDude commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterDoor.cs:
```diff
@@ -116,7 +115,7 @@ public override bool IsPassableFromInside(Vector3Int leavingTo, bool isServer, G
 			}
 
-			return !isClosed;
+			return true; //Should be able to walk out of closed doors
```
Does this affect one-sided doors?

As an aside, I wanted to take this passable stuff out of RegisterDoor and use DirectionalPassable instead.

Bod9001 replied:
I think it's a separate script for directional doors

Bod9001 replied:
return CheckViaDirectional(tryGetDir, direction); 
yeah, is done separately

NoooneyDude commented on UnityProject/Assets/Scripts/Player/Movement/PlayerMove.cs:
```diff
@@ -21,7 +21,7 @@ namespace Player.Movement
 	///     be walked into it.
 	/// </summary>
-	public class PlayerMove : NetworkBehaviour, IRightClickable, IServerSpawn, IActionGUI, ICheckedInteractable<ContextMenuApply>, RegisterPlayer.IControlPlayerState
+	public class OLDPlayerMove : NetworkBehaviour, IRightClickable, IServerSpawn, IActionGUI, ICheckedInteractable<ContextMenuApply>, RegisterPlayer.IControlPlayerState
```
Anything still use this? Can it be deleted?

Bod9001 replied:
let's see what explodes

Bod9001 committed:
Requested changes

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
Everything should use universal object position now

Bod9001 committed:
don't need this anymore

Bod9001 committed:
miscellaneous fixes

NoooneyDude commented on UnityProject/Assets/Scripts/Core/UniversalObjectPhysics.cs:
```diff
@@ -0,0 +1,1610 @@
+
+	//public const float DEFAULT_Friction = 9999999999999f;
+	public const float DEFAULT_Friction = 15f;
+	public const float DEFAULT_SLIDE_FRICTION = 9f;
```
Should these be serialised so different objects have different friction or are these some constants needed for some math

Bod9001 replied:
currently is constant, could be done per object basis,  could even be determined by tile to but Would cause issues with throwing Landing perfectly underneath mouse

NoooneyDude commented on UnityProject/Assets/Scripts/Messages/Server/PlayerMoveMessage.cs:
```diff
@@ -32,22 +32,22 @@ public override void Process(NetMessage msg)
+			// 	}
+			//
+			// 	ControlTabs.CheckTabClose();
+			// }
```
Can delete this script too?

NoooneyDude commented on UnityProject/Assets/Scripts/Objects/Construction/APCFrame.cs:
```diff
@@ -302,7 +302,7 @@ private void WrenchedStateInteraction(HandApply interaction)
 				econs.Pool();
 
-				GameObject WallMount = Spawn.ServerPrefab(APCObject, gameObject.GetComponent<CustomNetTransform>().ServerPosition, interaction.Performer.transform.parent, spawnItems: false).GameObject;
+				GameObject WallMount = Spawn.ServerPrefab(APCObject, gameObject.GetComponent<UniversalObjectPhysics>().transform.position, interaction.Performer.transform.parent, spawnItems: false).GameObject;
```
Don't need to get the component if we're just using it for `transform.position`.

NoooneyDude commented on UnityProject/Assets/Scripts/Managers/ComponentManager.cs:
```diff
@@ -0,0 +1,50 @@
+		}
+
+		return ObjectToPhysics[gameObject];
+	}
```
So as this could be used a lot we can probably do some little optimisations:

- use `ObjectToPhysics.TryGetValue()` for line `21` so we only do a dictionary lookup once (don't need a lookup on line `33`) (and need to return UOP after line `30` ofc).
- In similar fashion as `TryGetValue()` let's change this method so that it returns a boolean for if found or not, and uses an `out` parameter for the component (like the `TryGetComponent()`s), so that usages of this function don't need to check for null again (Unity makes it expensive).
Edit: actually you return an explicit null, not familiar enough if that changes things - probably not? I think semantically it is more consistent with other `Try` methods to do it as mentioned.

Bod9001 committed:
code improvements

Bod9001 committed:
Merge remote-tracking branch 'upstream/develop' into new-move

Bod9001 committed:
compile fix

Bod9001 committed:
fixes tests

NoooneyDude reviewed (approved):
Codacy has a few good suggestions but otherwise I'm happy 👍
